compiler/rustc_span/src/lev_distance.rs

```diff
@@ -1,49 +1,104 @@
-//! Levenshtein distances.
+//! Damerau-Levenshtein distances.
 //!
-//! The [Levenshtein distance] is a metric for measuring the difference between two strings.
+//! The [Damerau-Levenshtein distance] is a metric for measuring the difference between two strings.
+//! This implementation is a restricted version of the algorithm, as it does not permit modifying
+//! characters that have already been transposed.
 //!
-//! [Levenshtein distance]: https://en.wikipedia.org/wiki/Levenshtein_distance
+//! [Damerau-Levenshtein distance]: https://en.wikipedia.org/wiki/Damerau%E2%80%93Levenshtein_distance
 
 use crate::symbol::Symbol;
-use std::cmp;
+use std::{cmp, mem};
 
 #[cfg(test)]
 mod tests;
 
-/// Finds the Levenshtein distance between two strings.
+/// Finds the restricted Damerau-Levenshtein distance between two strings. Characters that have
+/// already been transposed may not be modified.
 ///
 /// Returns None if the distance exceeds the limit.
 pub fn lev_distance(a: &str, b: &str, limit: usize) -> Option<usize> {
-    let n = a.chars().count();
-    let m = b.chars().count();
-    let min_dist = if n < m { m - n } else { n - m };
+    let mut a = a.chars().collect::<Vec<_>>();
+    let mut b = b.chars().collect::<Vec<_>>();
+
+    // Ensure that `b` is the shorter string, minimizing memory use.
+    if a.len() < b.len() {
+        mem::swap(&mut a, &mut b);
+    }
 
+    let min_dist = a.len() - b.len();
+    // If we know the limit will be exceeded, we can return early.
     if min_dist > limit {
         return None;
     }
-    if n == 0 || m == 0 {
-        return (min_dist <= limit).then_some(min_dist);
-    }
 
-    let mut dcol: Vec<_> = (0..=m).collect();
+    // Strip common prefix.
+    loop {
+        match a.split_first().zip(b.split_first()) {
+            Some(((a_char, a_rest), (b_char, b_rest))) if a_char == b_char => {
+                a = a_rest.to_vec();
+                b = b_rest.to_vec();
```

Comment:
Declare a and b as slices to avoid allocation here, e.g.,:

```rust
    let a = a.chars().collect::<Vec<_>>();
    let mut a = a.as_slice();
```

Reply:
This could even use temporary lifetime extension
```rs
let mut a = &a.chars().collect::<Vec<_>>()[..];
```
though this might be considered less nice/idiomatic, depending on who you ask.

Reply:
I certainly felt like I was missing something here!